### PR TITLE
[Assignment Tracking] Trim assignments for untagged out of bounds stores

### LINKED ISSUE #65004: clang crashes on valid code at -O1 with debugging enabled (-g): Assertion `Traits::nonEmpty(a, b)' failed
This appears to be a recent regression.

Compiler Explorer: https://godbolt.org/z/rEGxzK7x3

```
[588] % clangtk -v
clang version 18.0.0 (https://github.com/llvm/llvm-project.git 012b148dd9402c6ad58805701863d31fb506caff)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda, version 11.8
[589] %
[589] % clangtk -O1 -g -c small.c
clangtk: /local/suz-local/software/clangbuild/llvm-project/llvm/include/llvm/ADT/IntervalMap.h:1187: bool llvm::IntervalMap<KeyT, ValT, N, Traits>::overlaps(KeyT, KeyT) const [with KeyT = unsigned int; ValT = unsigned int; unsigned int N = 16; Traits = llvm::IntervalMapHalfOpenInfo<unsigned int>]: Assertion `Traits::nonEmpty(a, b)' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clangtk -I/usr/local/include -O1 -g -c small.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'small.c'.
4.      Running pass 'Assignment Tracking Analysis' on function '@b'
 #0 0x00005628e445f800 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x3c71800)
 #1 0x00005628e445cc0f llvm::sys::RunSignalHandlers() (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x3c6ec0f)
 #2 0x00005628e43a7818 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007f04e500b520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007f04e505fa7c pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x96a7c)
 #5 0x00007f04e500b476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007f04e4ff17f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00007f04e4ff171b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #8 0x00007f04e5002e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
 #9 0x00005628e3c346e7 llvm::IntervalMap<unsigned int, unsigned int, 16u, llvm::IntervalMapHalfOpenInfo<unsigned int>>::overlaps(unsigned int, unsigned int) const (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x34466e7)
#10 0x00005628e3c45b8f (anonymous namespace)::MemLocFragmentFill::run(FunctionVarLocsBuilder*) AssignmentTrackingAnalysis.cpp:0:0
#11 0x00005628e3c503b0 llvm::AssignmentTrackingAnalysis::runOnFunction(llvm::Function&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x34623b0)
#12 0x00005628e3dc8b3e llvm::FPPassManager::runOnFunction(llvm::Function&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x35dab3e)
#13 0x00005628e3dc8d89 llvm::FPPassManager::runOnModule(llvm::Module&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x35dad89)
#14 0x00005628e3dc96c5 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x35db6c5)
#15 0x00005628e46c66e4 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x3ed86e4)
#16 0x00005628e56def9a clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x4ef0f9a)
#17 0x00005628e66c0db9 clang::ParseAST(clang::Sema&, bool, bool) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x5ed2db9)
#18 0x00005628e4f1a249 clang::FrontendAction::Execute() (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x472c249)
#19 0x00005628e4e999de clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x46ab9de)
#20 0x00005628e4fe9f1f clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x47fbf1f)
#21 0x00005628e181dc63 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x102fc63)
#22 0x00005628e1816b93 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#23 0x00005628e4ce395d void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#24 0x00005628e43a7ce0 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x3bb9ce0)
#25 0x00005628e4ce41de clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#26 0x00005628e4caa45a clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x44bc45a)
#27 0x00005628e4caaf2d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x44bcf2d)
#28 0x00005628e4cb67b4 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x44c87b4)
#29 0x00005628e181c140 clang_main(int, char**, llvm::ToolContext const&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x102e140)
#30 0x00005628e1731e33 main (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0xf43e33)
#31 0x00007f04e4ff2d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#32 0x00007f04e4ff2e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#33 0x00005628e18157d5 _start (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-18+0x10277d5)
clangtk: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 18.0.0 (https://github.com/llvm/llvm-project.git 012b148dd9402c6ad58805701863d31fb506caff)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
clangtk: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clangtk: note: diagnostic msg: /tmp/small-86de63.c
clangtk: note: diagnostic msg: /tmp/small-86de63.sh
clangtk: note: diagnostic msg:

********************
[590] % cat small.c
int a;
void b() {
  int c[2] = {0, 0};
  unsigned d = -1;
  if (a)
    c[a] = c[d] &= a;
  b();
}
```

### PR DESCRIPTION
**Pull request by OCHyams**
Fixes #65004 by trimming assignments from out of bounds stores (out of bounds
of either the base variable or the backing alloca). If there's no overlap at
all or the out of bounds access starts at a negative offset from the alloca,
the assignment is simply skipped. IMO it's not worth the extra code to
trim the negative-alloca-offset-but-overlaps-variable case.

### CODE REVIEW

**Review by SLTozer — COMMENTED** [review]
Largely LGTM, a couple minor comments inline.

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp [hunk 1]
```diff
@@ -1979,20 +1979,22 @@ static AssignmentTrackingLowering::OverlapMap buildOverlapMapAndRecordDeclares(
...
+
+          // Skip this assignment if the affected bits are outside of the
+          // variable fragment.
+          if (!at::calculateFragmentIntersect(
+                  I.getModule()->getDataLayout(), Info->Base,
+                  Info->OffsetInBits, Info->SizeInBits, DAI, FragInfo) ||
+              (FragInfo && FragInfo->SizeInBits == 0))
+            continue;
+
+          // FragInfo from calculateFragmentIntersect is is nullopt if the
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp:1992]: done

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp [hunk 2]
```diff
@@ -1979,20 +1979,22 @@ static AssignmentTrackingLowering::OverlapMap buildOverlapMapAndRecordDeclares(
...
+          // resultant fragment matches DAI's fragment or entire variable - in
+          // that case just copy the fragment from DAI. Now nullopt means "no
+          // fragment info", as per usual.
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp]: Nope - hopefully the new wording improves clarity?

File: llvm/test/DebugInfo/assignment-tracking/X86/untagged-store-assignment-extra-checks.ll [hunk 3]
```diff
@@ -0,0 +1,94 @@
+; RUN: llc %s -stop-after=finalize-isel -o - \
+; RUN: | FileCheck %s
+;--implicit-check-not=DBG_
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/test/DebugInfo/assignment-tracking/X86/untagged-store-assignment-extra-checks.ll]: done

File: llvm/test/DebugInfo/assignment-tracking/X86/untagged-store-assignment-extra-checks.ll [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+declare dso_local void @a(i32)
+
+define dso_local void @b() local_unnamed_addr !dbg !14 {
+entry:
+  %c = alloca [4 x i16], align 8, !DIAssignID !24
+  %arrayidx = getelementptr inbounds [4 x i16], ptr %c, i64 0, i64 2
+  call void @llvm.dbg.assign(metadata i1 undef, metadata !18, metadata !DIExpression(DW_OP_LLVM_fragment, 128, 32), metadata !24, metadata ptr %arrayidx, metadata !DIExpression()), !dbg !26
+
+;; Set variable value (use a call to prevent eliminating redundant DBG_VALUEs).
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply llvm/test/DebugInfo/assignment-tracking/X86/untagged-store-assignment-extra-checks.ll:18]: Ah, no the comment is just copy-pasted. The redundant elimination is fine here as it doesn't impact the behaviour we're testing. The initial dbg.assign is needed to associate a variable (fragment) with the base storage, but we don't need to see a DBG_VALUE proving that.

**Review by OCHyams — COMMENTED** [review]
Thanks! I've addressed your review comments.

File: llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SLTozer) [reply llvm/lib/CodeGen/AssignmentTrackingAnalysis.cpp]: It does, that makes more sense!

**Review by SLTozer — APPROVED** [review]

**Pull request merged**